File: lib/audit.ts
import { prisma } from "@/lib/prisma";
import type { Prisma } from "@prisma/client";
import { log } from "@/lib/logger";

/**
 * Fire-and-forget audit log writer.
 * Never throws — audit failures must not break the request.
 */
export async function auditLog(params: {
  userId: string;
  orgId?: string | null;
  action: string;
  targetType: string;
  targetId?: string | null;
  metadata?: Record<string, unknown>;
}): Promise<void> {
  try {
    await prisma.auditLog.create({
      data: {
        userId: params.userId,
        orgId: params.orgId ?? null,
        action: params.action,
        targetType: params.targetType,
        targetId: params.targetId ?? null,
        metadata: (params.metadata ?? undefined) as Prisma.InputJsonValue | undefined,
      },
    });
  } catch (e) {
    log.error("Failed to write audit log", {
      component: "audit",
      action: params.action,
      targetType: params.targetType,
      err: e instanceof Error ? e.message : String(e),
    });
  }
}

File: guardian-engine/src/helpers/audit.rs
use serde_json::Value;
use sqlx::PgPool;

/// Fire-and-forget audit log writer. Never panics.
pub async fn audit_log(
    pool: &PgPool,
    user_id: &str,
    org_id: Option<&str>,
    action: &str,
    target_type: &str,
    target_id: Option<&str>,
    metadata: Option<Value>,
) {
    let id = cuid2::create_id();
    let _ = sqlx::query(
        r#"INSERT INTO "AuditLog" (id, "userId", "orgId", action, "targetType", "targetId", metadata, "createdAt")
           VALUES ($1, $2, $3, $4, $5, $6, $7, NOW())"#
    )
    .bind(&id)
    .bind(user_id)
    .bind(org_id)
    .bind(action)
    .bind(target_type)
    .bind(target_id)
    .bind(metadata)
    .execute(pool)
    .await;
}
